src/main/java/ch/njol/skript/sections/EffSecSpawn.java

```diff
@@ -0,0 +1,165 @@
+
+	public static class SpawnEvent extends Event {
+		private final Entity entity;
+
+		public SpawnEvent(Entity entity) {
+			this.entity = entity;
+		}
+
+		public Entity getEntity() {
+			return entity;
+		}
+
+		@Override
+		public @NotNull HandlerList getHandlers() {
+			throw new IllegalStateException();
+		}
+	}
+
+	static {
+		Skript.registerSection(EffSecSpawn.class,
+			"(spawn|summon) %entitytypes% [%directions% %locations%]",
+			"(spawn|summon) %number% of %entitytypes% [%directions% %locations%]");
+		EventValues.registerEventValue(SpawnEvent.class, Entity.class, new Getter<Entity, SpawnEvent>() {
+			@Override
+			public Entity get(SpawnEvent spawnEvent) {
+				return spawnEvent.getEntity();
+			}
+		}, 0);
+	}
+
+	@Nullable
+	public static Entity lastSpawned = null;
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Location> locations;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<EntityType> types;
+	@Nullable
+	private Expression<Number> amount;
+
+	@Nullable
+	private Trigger trigger;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs,
+						int matchedPattern,
+						Kleenean isDelayed,
+						ParseResult parseResult,
+						@Nullable SectionNode sectionNode,
+						@Nullable List<TriggerItem> triggerItems) {
+		amount = matchedPattern == 0 ? null : (Expression<Number>) (exprs[0]);
+		types = (Expression<EntityType>) exprs[matchedPattern];
+		locations = Direction.combine((Expression<? extends Direction>) exprs[1 + matchedPattern], (Expression<? extends Location>) exprs[2 + matchedPattern]);
+
+		if (sectionNode != null)
+			trigger = loadCode(sectionNode, "spawn", SpawnEvent.class);
+
+		return true;
+	}
+
+	@Override
+	@Nullable
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	protected TriggerItem walk(Event e) {
+		lastSpawned = null;
+
+		Object localVars = Variables.copyLocalVariables(e);
+
+		Consumer<? extends Entity> consumer;
+		if (trigger != null) {
+			consumer = o -> {
+				SpawnEvent spawnEvent = new SpawnEvent(o);
+				// Copy the local variables from the calling code to this section
+				Variables.setLocalVariables(spawnEvent, localVars);
+				trigger.execute(spawnEvent);
+			};
+		} else {
+			consumer = null;
+		}
+
+		Number a = amount != null ? amount.getSingle(e) : 1;
+		if (a != null) {
+			EntityType[] ts = types.getArray(e);
+			for (Location l : locations.getArray(e)) {
+				for (EntityType type : ts) {
+					for (int i = 0; i < a.doubleValue() * type.getAmount(); i++) {
+						if (consumer != null)
+							lastSpawned = type.data.spawn(l, (Consumer) consumer);
+						else
+							lastSpawned = type.data.spawn(l);
+					}
+				}
+			}
+		}
+
+		return super.walk(e, false);
+	}
+
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return "spawn " + (amount != null ? amount.toString(e, debug) + " of " : "") + types.toString(e, debug) + " " + locations.toString(e, debug);
```

Comment:
Event can be null. When I was doing separate syntax toString in my own addons, I was running into null pointer exception when I didn't handle if the event was null. I usually sent like:
```
if (e == null)
    return "spawn entity";
```
Does this still happen if the verbose level is set to `debug`?

Reply:
This method overrides `Debuggable#toString(@Nullable Event, boolean)`, and the only thing it does with the `Event` argument, is pass it onto other `Debuggable#toString(@Nullable Event, boolean)` methods, so this isn't an issue.

The toString method is supposed to look like the line that the syntax element matched, so the expression `{_x}`'s toString will return `"{_x}"` and so forth. The debug parameter is usually false (IIRC it's only true for getting the current TriggerItem in exception printing, or in some cases when verbosity is debug), and the event will always be null when debug is false (see the method's Javadocs). 

Most effects and expressions don't need to have special case scenarios for when debug is true, except for fundamental expressions such as Variable or ConvertedExpression.